### PR TITLE
fix: Make payments recoverable

### PR DESCRIPTION
**Pull request by erights**
Associate with each purse a recover set of live payments that can be recovered in emergencies, such as when a payment seems lost or stuck. Recovering a payment is like cancelling an outstanding check.

Withdrawing from a purse puts the resulting payment in that purse's recovery set.

### CODE REVIEW

File: packages/ERTP/src/paymentLedger.js
Comment (michaelfig) [inline packages/ERTP/src/paymentLedger.js]: Is protecting this `.add` with an if statement useful?  It looks like just a distraction.

File: packages/ERTP/src/paymentLedger.js
Comment (michaelfig) [inline packages/ERTP/src/paymentLedger.js]: Same here?

File: packages/ERTP/src/types.js
Comment (michaelfig) [inline packages/ERTP/src/types.js]:
```suggestion
 * cancellation is acknowledged, you can spend the assets at stake on other
```

File: packages/ERTP/test/unitTests/test-recovery.js
Comment (michaelfig) [inline packages/ERTP/test/unitTests/test-recovery.js]: Do you also want to assert emptySet on `bobPurse` before and after depositing `payment2`?

**Review by michaelfig — COMMENTED** [review]
Looking good so far.  Let me know when you're ready for a final review (especially if you plan on doing something about `virtual-purse.js`, which notably the `vbank` relies on).

**Comment by erights** [comment]
> Looking good so far. Let me know when you're ready for a final review (especially if you plan on doing something about virtual-purse.js, which notably the vbank relies on).

I took a look at vbank again and decided not to do it in this PR. Would you like to do vbank after this PR?

With vbank postponed, this is now ready for review. Thanks.

**Comment by michaelfig** [comment]
> With vbank postponed, this is now ready for review. Thanks.

It seems I was only tagged as a reviewer because of my codeowner status for `virtual-purse.js`.  I would prefer if somebody else took responsibility for this review, as I may be lacking some necessary context.

**Comment by erights** [comment]
With @Chris-Hibbert OOO, @dtribble can you review? Thanks.

File: packages/ERTP/src/types.js
Comment (dtribble) [inline packages/ERTP/src/types.js:354]: this does make purses non-robust servers in that any holder of a purse can impair the use by any other holder of a purse. I think that's OK , but it is a change.

File: packages/vats/src/virtual-purse.js
Comment (dtribble) [inline packages/vats/src/virtual-purse.js]: We shouldn't commit with this.  At the very least there would need to be an issue, but what are the design issues for this?

File: packages/ERTP/src/paymentLedger.js
Comment (dtribble) [inline packages/ERTP/src/paymentLedger.js]: This needs documentation here. It's too clever to leave without explanation.

File: packages/ERTP/src/paymentLedger.js
Comment (dtribble) [inline packages/ERTP/src/paymentLedger.js]: Keep this code somewhere, but all the claim operations in our code are in tests. Because of Zoe, all real usage ends up depositing into Zoe escrow. So all the `claim` and related operations can be removed, leaving just the recovery logic around the `withdraw` operation.

File: packages/ERTP/src/purse.js
Comment (dtribble) [inline packages/ERTP/src/purse.js]: As above, delete

File: packages/ERTP/src/types.js
Comment (dtribble) [inline packages/ERTP/src/types.js]: Delete.

File: packages/vats/src/virtual-purse.js
Comment (dtribble) [inline packages/vats/src/virtual-purse.js]: delete

File: packages/vats/src/virtual-purse.js
Comment (dtribble) [inline packages/vats/src/virtual-purse.js]: This should be implemented. It looks to me like it's exactly the same code; all the recovery is on the JS side.

**Review by dtribble — CHANGES_REQUESTED** [review]
Key things to complete:
- remove the `claim operations`
- add an implementation comment
- provide the implementation for virtual purses too

File: packages/ERTP/src/types.js
Comment (erights) [reply packages/ERTP/src/types.js:354]: (we discussed and agreed it is ok. Because anyone holding a purse can withdraw everything from it, we encourage a programming pattern where a single purse is never help my multiple mutually suspicious parties, or even by multiple non-coordinated parties. Within that framing, we are not introducing any new vulnerabilities.)

File: packages/ERTP/src/paymentLedger.js
Comment (erights) [reply packages/ERTP/src/paymentLedger.js]: Git keeps the code. I really doubt we will need it again. It is sufficiently easy to reproduce anyway that I'm just "deleting" it.

File: packages/vats/src/virtual-purse.js
Comment (erights) [reply packages/vats/src/virtual-purse.js]: Since you have a separate review comment on virtual-purse `getRecoverySet` below and I've already deleted `claim`, I'm resolving this.

File: packages/vats/src/virtual-purse.js
Comment (erights) [reply packages/vats/src/virtual-purse.js]: What I'm confused about here is whether a vpurse
   * a vpurse should forward to an underlying ERTP purse which implements this logic, or 
   * should reimplement that on its own.
   * one case of vpurse should forward and the other case should reimplement.

I suspect the third bullet is the right answer, but I'm not well enough oriented to be confident.

@michaelfig can you do the vpurse adjustment and I'll review?

@dtribble if yes, would you rather @michaelfig do that in a separate PR or as an enhancement to this PR?

Either way, when this PR is ready, someone else should merge it if I don't seem to be available.

Comment (erights) [reply packages/vats/src/virtual-purse.js]: Thanks

Comment (michaelfig) [reply packages/vats/src/virtual-purse.js]: > - should reimplement that on its own

In our uses of this `virtual-purse.js`, each vpurse should implement its own recoverySet, so that recoverAll calls vpurse.deposit (which is where the vpurse magic happens).  I think the right thing is to make it part of the current PR.  I can attempt this if you don't have bandwidth for now.

Note that vpurses are not yet Virtual nor Durable objects, and even without that restriction, are somewhat simpler than `@agoric/ertp` purses.  That makes things easier.

Comment (erights) [reply packages/vats/src/virtual-purse.js]: >  I can attempt this if you don't have bandwidth for now.

Yes please! Much appreciated!

Comment (michaelfig) [reply packages/vats/src/virtual-purse.js]: Oh, when implementing I found that this is actually a case of "forwarding to an ERTP purse which implements this logic."  But it's still pretty simple.

File: packages/vats/src/virtual-purse.js
Comment (michaelfig) [inline packages/vats/src/virtual-purse.js:61]: Ironically, I could have used `E(recoveryPurse).claim` here.

**Comment by erights** [comment]
@michaelfig , please review my changes to your changes at https://github.com/Agoric/agoric-sdk/pull/5008/commits/5d71a1cdea4bc8bf4c4e0de07488f520429eaea0

**Review by erights — COMMENTED** [review]
@michaelfig FWIW, your changes at https://github.com/Agoric/agoric-sdk/pull/5008/commits/b337db7d050faed61311a329bf45361e11505eb9 LGTM modulo my refinement in the next commit.

File: packages/ERTP/src/purse.js
Comment (dtribble) [inline packages/ERTP/src/purse.js:70]: ha!  Glad to see that show up. It's also noteworthy that the deletion is indirect because of the operation invoked. It would be more complicated if the iteration had to decide whether to delete.

File: packages/ERTP/src/paymentLedger.js
Comment (dtribble) [inline packages/ERTP/src/paymentLedger.js:61]: love it.

File: packages/vats/src/virtual-purse.js
Comment (dtribble) [inline packages/vats/src/virtual-purse.js]: If you are just going to burn, is there any reason to have a transition to a recoverable claim?

File: packages/vats/src/virtual-purse.js
Comment (dtribble) [inline packages/vats/src/virtual-purse.js:75]: This is an interesting reason to have a `claim` method on purse:  minting!

Comment (dtribble) [reply packages/vats/src/virtual-purse.js:75]: (could be added separately.

**Review by dtribble — APPROVED** [review]
A few comments and questions on the virtual-purse, but it all looks correct.

File: packages/vats/src/virtual-purse.js
Comment (michaelfig) [reply packages/vats/src/virtual-purse.js]: Yeah, I was thinking about this last night.  No need for this recoverableClaim, nor for the other retain (in which we just deposit to the escrow purse).

**Pull request merged**
